src/conduit-lwt-unix/conduit_lwt_unix.ml

```diff
@@ -264,6 +264,15 @@ let connect_with_tls_native ~ctx (`Hostname hostname, `IP ip, `Port port) =
       Conduit_lwt_tls.X509.private_of_pems ~cert ~priv_key
       >|= fun certificate -> Some (`Single certificate))
   >>= fun certificates ->
+  let hostname =
+    try Domain_name.(host_exn (of_string_exn hostname)) with
+      Invalid_argument msg ->
+      let s =
+        Printf.sprintf "couldn't convert %s to a [`host] Domain_name.t: %s"
+          hostname msg
+      in
+      invalid_arg s
+  in
```

Comment:
You probably should update `Conduit_lwt_unix_ssl.real.ml{,i}` and use `Domain_name.to_string` in this place instead of raising an error. I think it's fine to break the API (and require an `[ host ] Domain_name.t` instead of a simple `string` as the host) and be more restrictive even if we want to use OpenSSL. WDYT @avsm?

Reply:
As explained above, I'd leave the API change for someone in the future. I still don't have any idea what "The `conduit` library takes care of establishing and listening for TCP and SSL/TLS connections for the Lwt and Async libraries." means in respect to error handling. Code-wise since `connect` raises, I would find it obscure if only the Domain_name.host / Domain_name.of_string would result in an error (instead I'd expect either "no exceptions are raised" or "no result type is used") -- and the current code is "no result type is used, exceptions are raised"